### ISSUE TITLE
Regression causing some requests to block in 0.94.0

### ISSUE THREAD

**Issue by mloughran** [excon/excon#799]
I'm observing that following the update from 0.93.0 to 0.94.0 the response processing becomes blocked, resulting in an excon timeout being raised eventually. I haven't had time yet to narrow down the scenario or bisect changes, but am creating this in case others are also seeing issues. These are the facts:

* Request for 1.5MB file from a GCP cloud storage bucket, from within GCP, within docker container
* Works on 0.93.0, fails with excon timeout on 0.94.0
* Packet traces captured with excon 0.93.0 and 0.94.0 show that all/most of the data was received by 0.94.0 (the `pcap` files are about the same size). There are some differences  – the 0.94.0 trace doesn't include any TCP window full/update messages – but I don't know if this are significant. I'm happy to share `pcap` files privately if helpful for debugging.
* Failure is totally reproducable.
* Downloading a small text file always works.

**Comment by geemus** [comment]
@mloughran - Thanks for the heads up. Will see what we can do.

@vasi-stripe - Any thoughts on this? Since you did most of the related changes I was hoping you might have some insights given this scenario. Thanks!

**Comment by vasi-stripe** [comment]
Hmm, we did notice some anomalies in the first stages of roll-out, but nothing reproducible. And then I got pulled onto other priorities.

@mloughran Can you share more info about how to repro this?

**Comment by mloughran** [comment]
Thanks both. I've now replicated the issue using a simple unauthenticated GET to a GCP cloud storage bucket, however it only appears to happen when running within the GCP network. Running git bisect, the issue was introduced by 7ac3568.

The following blocks on a GCP VM (have tried 2 hosts), but not on my Mac or on another random linux host. It seems somewhat dependant on the filesize, but `index.json` is just a random JSON file from my HD which was roughly the same size, so it's not _that_ sensitive.

```ruby
require "excon"

resp = Excon.get(
  "https://storage.googleapis.com/storage/v1/b/test-excon-bucket/o/index.json",
  query: { alt: "media" },
)

puts resp.status
```

Interrupting a frozen call to `Excon.get` results in the following stacktrace

	Traceback (most recent call last):
		9: from test.rb:7:in `<main>'
		8: from /var/lib/gems/2.7.0/gems/excon-0.94.0/lib/excon.rb:250:in `get'
		7: from /var/lib/gems/2.7.0/gems/excon-0.94.0/lib/excon/connection.rb:287:in `request'
		6: from /var/lib/gems/2.7.0/gems/excon-0.94.0/lib/excon/connection.rb:456:in `response'
		5: from /var/lib/gems/2.7.0/gems/excon-0.94.0/lib/excon/middlewares/response_parser.rb:7:in `response_call'
		4: from /var/lib/gems/2.7.0/gems/excon-0.94.0/lib/excon/response.rb:159:in `parse'
		3: from /var/lib/gems/2.7.0/gems/excon-0.94.0/lib/excon/socket.rb:58:in `read'
		2: from /var/lib/gems/2.7.0/gems/excon-0.94.0/lib/excon/socket.rb:198:in `read_nonblock'
		1: from /usr/lib/ruby/2.7.0/openssl/buffering.rb:182:in `read_nonblock'
	/usr/lib/ruby/2.7.0/openssl/buffering.rb:182:in `sysread_nonblock': read would block (OpenSSL::SSL::SSLErrorWaitReadable)
		10: from test.rb:7:in `<main>'
		 9: from /var/lib/gems/2.7.0/gems/excon-0.94.0/lib/excon.rb:250:in `get'
		 8: from /var/lib/gems/2.7.0/gems/excon-0.94.0/lib/excon/connection.rb:287:in `request'
		 7: from /var/lib/gems/2.7.0/gems/excon-0.94.0/lib/excon/connection.rb:456:in `response'
		 6: from /var/lib/gems/2.7.0/gems/excon-0.94.0/lib/excon/middlewares/response_parser.rb:7:in `response_call'
		 5: from /var/lib/gems/2.7.0/gems/excon-0.94.0/lib/excon/response.rb:159:in `parse'
		 4: from /var/lib/gems/2.7.0/gems/excon-0.94.0/lib/excon/socket.rb:58:in `read'
		 3: from /var/lib/gems/2.7.0/gems/excon-0.94.0/lib/excon/socket.rb:196:in `read_nonblock'
		 2: from /var/lib/gems/2.7.0/gems/excon-0.94.0/lib/excon/socket.rb:203:in `rescue in read_nonblock'
		 1: from /var/lib/gems/2.7.0/gems/excon-0.94.0/lib/excon/socket.rb:297:in `select_with_timeout'
	/var/lib/gems/2.7.0/gems/excon-0.94.0/lib/excon/socket.rb:297:in `select': Interrupt

Hope the above helps @vasi-stripe.

**Comment by geemus** [comment]
@mloughran I have a suspicion that this commit may help, but it's been hard to know since I haven't been able to reproduce: https://github.com/excon/excon/commit/912045d0b91731520d68118f0a5ffce2a16aa713. Would love any input or confirmation you could offer.

**Comment by geemus** [comment]
P.S. The other commits on master should be safe/unrelated, so you should be able to test from there (rather than needing to cherry pick or something).

**Comment by mloughran** [comment]
@geemus bingo, I can confirm that 912045d0b91731520d68118f0a5ffce2a16aa713 does indeed fix the issue – sorry been very busy this end and it's taken a few days to try this out. Thanks very much for digging into this and I hope you've all recovered from covid by now.

**Comment by geemus** [comment]
@mloughran Awesome, thanks. No worries, I appreciate your help and patience. At least mostly recovered, though some lingering fatigue.

The fix was in the part that seemed most suspicious, but it's still a little unsatisfying that I can't quite put my finger on a reproduction case. In any event, I'm glad to hear this seems to have done the trick and I'll plan on a release early next week (I don't like to release on weekends when I tend to be checking in less, just in case). I'll plan to update this when I do that as well so that you will be aware. Thanks!

**Comment by geemus** [comment]
I've released v0.95.0 this morning, I believe this should now be resolved, but just let me know if you see any further issues.
